pkg/basinprovider/provider.go

```diff
@@ -132,14 +135,79 @@ func (bp *BasinProvider) Push(ctx context.Context, ns string, rel string, tx bas
 	return err
 }
 
+// Upload uploads a file to th server.
+func (bp *BasinProvider) Upload(
+	ctx context.Context, ns string, rel string, r io.Reader, signer *app.Signer, progress io.Writer,
+) error {
+	uploadFuture, uploadRelease := bp.p.Upload(ctx, func(p Publications_upload_Params) error {
+		if err := p.SetNs(ns); err != nil {
+			return fmt.Errorf("setting sig: %s", err)
+		}
+
+		if err := p.SetRel(rel); err != nil {
+			return fmt.Errorf("setting sig: %s", err)
+		}
+
+		return nil
+	})
+	defer uploadRelease()
+
+	callback := uploadFuture.Callback()
+
+	buf := make([]byte, 0, 4*1024)
+	for {
+		n, err := r.Read(buf[:cap(buf)])
+		buf = buf[:n]
+		if err != nil {
+			if err != io.EOF {
+				return fmt.Errorf("read file: %s", err)
+			}
+			break
+		}
+		signer.Sum(buf)
+
+		err = callback.Write(ctx, func(p Publications_Callback_write_Params) error {
+			return p.SetChunk(buf)
+		})
+		if err != nil {
+			return fmt.Errorf("write: %s", err)
+		}
+
+		_, _ = progress.Write(buf)
+	}
+
+	doneFuture, doneRelease := callback.Done(ctx, func(p Publications_Callback_done_Params) error {
```

Comment:
calls done sending the signature, and wait for completion

Reply:
so the difference with the "WAL"  message workflow is that, here, there is one signature per file whereas in the other workflow, each WAL message was signed?

Reply:
yeah! exactly